lib/linux_admin/network_interface.rb

```diff
@@ -0,0 +1,138 @@
+
+      cidr_ip = parse_ip_output(ip_output, /inet/, 1)
+      cidr_ip.split('/')[0] if cidr_ip
+    end
+
+    # Retrieve the IPv6 address assigned to the interface
+    #
+    # @return [String] IPv6 address for the managed interface
+    def address6(scope = :global)
+      return nil unless (ip_output = ip_show)
+
+      ip_regex = /inet6 .* scope #{scope}/
+      cidr_ip = parse_ip_output(ip_output, ip_regex, 1)
+      cidr_ip.split('/')[0] if cidr_ip
+    end
+
+    # Retrieve the MAC address associated with the interface
+    #
+    # @return [String] the MAC address
+    def mac_address
+      return nil unless (ip_output = ip_show)
+      parse_ip_output(ip_output, %r{link/ether}, 1)
+    end
+
+    # Retrieve the IPv4 sub-net mask assigned to the interface
+    #
+    # @return [String] IPv4 netmask
+    def netmask
+      return nil unless (ip_output = ip_show)
+
+      cidr_ip = parse_ip_output(ip_output, /inet/, 1)
+      IPAddr.new('255.255.255.255').mask(cidr_ip.split('/')[1]).to_s if cidr_ip
+    end
+
+    # Retrieve the IPv6 sub-net mask assigned to the interface
+    #
+    # @return [String] IPv6 netmask
+    def netmask6(scope = :global)
+      return nil unless (ip_output = ip_show)
+
+      ip_regex = /inet6 .* scope #{scope}/
+      cidr_ip = parse_ip_output(ip_output, ip_regex, 1)
+      IPAddr.new('ffff:ffff:ffff:ffff:ffff:ffff:ffff:ffff').mask(cidr_ip.split('/')[1]).to_s if cidr_ip
+    end
+
+    # Retrieve the IPv4 default gateway associated with the interface
+    #
+    # @return [String] IPv4 gateway address
+    def gateway
+      result = run(cmd("ip"), :params => ["route"])
+      return nil if result.failure?
+
+      parse_ip_output(result.output, /^default/, 2)
+    end
+
+    private
+
+    # Detects the local linux distro and returns the associated NetworkInterface subclass
+    #
+    # @return [Class] The NetworkInterface subclass associated with the local distro
+    def self.dist_class_uncached
+      if [Distros.rhel, Distros.fedora].include?(Distros.local)
+        return NetworkInterfaceRH
+      else
+        return NetworkInterfaceGeneric
+      end
+      raise "No matching class for distro #{Distros.local}"
+    end
+    private_class_method :dist_class_uncached
+
+    # Parses the output of `ip addr show`
+    #
+    # @param output [String] The command output
+    # @param regex  [Regexp] Regular expression to match the desired output line
+    # @param col    [FixNum] The whitespace delimited column to be returned
+    # @return [String] The parsed data
+    def parse_ip_output(output, regex, col)
+      the_line = output.split("\n").detect { |l| l =~ regex }
+      the_line.nil? ? nil : the_line.strip.split(' ')[col]
+    end
+
+    # Runs the command `ip addr show <interface>`
+    #
+    # @return [String] The command output, nil on failure
+    def ip_show
+      result = run(cmd("ip"), :params => ["addr", "show", @interface])
+      result.success? ? result.output : nil
```

Comment:
It would definitely fail if they initialized with an interface name that doesn't exist, but we are not really giving good feedback for that situation.  Maybe raising there would be better anyway?  Would we want to implement our own exception class in that case and see if we can figure out if that is what went wrong?  After giving this another look I do want to change something about it because right we are not giving any feedback at all.